### ISSUE TITLE
Detecting tiny objects

### ISSUE THREAD

**Issue by thealchemist89** [yhenon/keras-frcnn#120]
Hi everyone,

I was reading about how the connection of the anchor boxes to each of the spatial dimension of the feature map from the convolutional layer of the network affects the recall rate of the detector. This is in reference to this paper: [https://www.merl.com/publications/docs/TR2016-144.pdf](url), page 9/18, Section 3.1 - Small Proposal Generation

In this repository:

1) Does any one knows how we can <modify> the attachment of our anchor boxes to the desired feature map (example, conv2_2 layer/conv3_3 layer/conv4_3 layer)?

2) Does any one played around with what is the best feature map for our anchor boxes to be attached to for ResNet50?

Thanks everyone for reading :)

**Comment by shuuchen** [comment]
I think that modifying self.anchor_box_scales to small sizes in config.py will work.

**Comment by thealchemist89** [comment]
If I may ask, how do you specify the use of conv4_3 or it is already in this configuration? I was probing into resnet.py

**Comment by olaff09** [comment]
Hi @thealchemist89, seems like we're solving similar problems, I read that paper too. I was going to try that multi-scale implementation too, but I haven't had time. There are many ways of doing it, from the top of my head I was thinking:

Separating the base layers function into 3, each returning the desired layer that you want  (conv2_2 layer/conv3_3 layer/conv4_3 layer), then concatenating the 3 outputs making them match their size, and pass the concatenated result to the rpn.

There are many techniques that I am thinking to apply, and I will surely share the code once tested.

The main issue is that the output of the base layers have a stride of 16pixels, this means that if your objects are smaller, or close to that size in pixels, they can be missed (this is some of my problem). A cheap way of improving your recognition is to double the size of the images before they enter the neural network, this will come with added computational cost, but almost any change that you apply to find smaller objects will likely make it heavier.

**Comment by thealchemist89** [comment]
Hi @olaff09 ,

Thank you for your valuable input. Hmm, I looked through the codes, it seems that the input image will be resized to at least 600 pixels if it falls below. Are you indicating that if I have a 480p input image, I should inject it with 720p as an example? 

Thank you for your advice!

**Comment by shuuchen** [comment]
@thealchemist89 
Just modify self.im_size in config.py

**Comment by olaff09** [comment]
@thealchemist89, 
You can change the im_size as @shuuchen suggests, but I wanted to maintain the exact aspect ratio, so I overwrote the resize function to be 2*width and 2*height.

The idea is that your objects in the resized image should be bigger than 16pixels (width and height) to increase the accuracy, so you should adjust the resizing to match this requirement. Of course you should keep in mind the distortion that resize does and some other stuff.

**Comment by thealchemist89** [comment]
@olaff09 ,

Thanks for your inputs, much appreciated. May I ask, did you manage to modify the code to concatenate the neural network for small/medium/large objects?

**Comment by olaff09** [comment]
Sure, we're all learning here :)

I haven't had time to develop this, since I am doing this research mostly on my free time, but I will surely do it sometime this month.

Here are some other papers on this matter if you'd like to start with it, this approach is called "multi-scale"

https://arxiv.org/abs/1512.04143
https://arxiv.org/abs/1606.05413

**Comment by thealchemist89** [comment]
@olaff09,

Sounds good, fellow hobbyist. I will implement the concatenation soon as well. Update you as I come along! I was reading this paper though:

https://arxiv.org/abs/1706.02863

**Comment by shuuchen** [comment]
I 'd rather recommend SSD for multi-scale recognition, which is as accurate and faster.
https://arxiv.org/abs/1512.02325

**Comment by olaff09** [comment]
@thealchemist89 thank you for that paper, a nice read, much easier to understand, hoping to hear from your updates!

@shuuchen Thanks, I've looked at SSD but haven't had time to get to understand the box proposal method, depending on how faster it can be, I might switch since the processing I am doing is pretty intensive (High-res images)

**Comment by olaff09** [comment]
What worries me about SSD is it's performance on small objects, according to [this ](https://arxiv.org/abs/1611.10012) paper, which extensively studied FRCNN vs SSD vs RFCN with multiple architectures, leaves FRCNN as the most accurate, but a bit slower. **Attaching graphic below**

For my purposes, I need the most accuracy I can get.

![image](https://user-images.githubusercontent.com/8171261/29088271-d8d35842-7c35-11e7-8fe3-eacc7f560bec.png)
